core-java/src/main/java/eu/neverblink/jelly/core/utils/IteratorUtils.java

```diff
@@ -0,0 +1,119 @@
+package eu.neverblink.jelly.core.utils;
+
+import java.util.Iterator;
+import java.util.NoSuchElementException;
+import java.util.function.Function;
+
+/**
+ * Utility class for working with iterators.
+ * <p>
+ * This class provides methods to transform and manipulate iterators in a functional style.
+ */
+public class IteratorUtils {
```

Comment:
Please make final... just in case. And add a comment to javadoc that it's intended for implementers of the adapters.